### ISSUE TITLE
Sealed classes

### ISSUE THREAD

**Issue by CoskunSunali** [linq2db/linq2db#1281]
I see that some of the classes are marked with the `sealed` keyword while others are not.

From extensibility point of view, sealing classes causes problems.

Is there a specific reason for sealing classes or would you mind removing the seals or accepting a PR?

Non-sealed: `LinqToDB.Mapping.TableAttribute`, `LinqToDB.Mapping.IdentityAttribute`,  `LinqToDB.Mapping.NullableAttribute`

Sealed: `LinqToDB.Mapping.PrimaryKeyAttribute`

**Comment by sdanyliv** [comment]
Not sure that i understand why do you need extending such functionlity. It’s just declarative attributes and information may be cached.

**Comment by sdanyliv** [comment]
As i know when we read these attributes we expect exact type. Will check.
And just for clarification. For linq2db it is only declarative information for generating correct SQL and mappers, no additional validations are performed. If you need this validation you can perform this by handling SQL AST before it is converted SQL and sent to the server.

If you need row level security i suggest to write extensions (just sample no pets were harmed)
```cs
public static IQueryable<Customer> GetAllowedCustomers(this IDataContext db)
{
     var query = db.GetTable<Customer>().AsQueryable();
     If (!CurrentUser.IsAdministrator)
     {
          query = query.Where(c => c.IsActive);
          query = 
              from c in query
              from p in db.GetTable<UserPermissions>().InnerJoin(p => p.UserId == c.UserId == CurrentUser.UserId 
                 && p.CustomerGroup == c.CustomerGroup && p.IsActive)
              select c;
     }
     return query;
}
```
Then you can use query returned by this function anywhere in linq queries and be sure that security is strong and no-one can select non related data. Then linq2db will optimize query decomposition that we have introduced and execute optimized SQL.

For sure you can write these extension based on attributes - you can read attributes and apply own filters.

**Comment by jogibear9988** [comment]
It is also possible to implement your own Metadata Reader, wich could support your attributes.

Or you could use own Attributes and Apply Fluent Mapping by reading your own attributes via reflection

**Comment by jogibear9988** [comment]
@sdanyliv maybe we should look that all attributes are sealed or all are not!

**Comment by CoskunSunali** [comment]
> just sample no pets were harmed

😃 

I have only tried it with the `TableAttribute` and it worked fine. So I thought you don't check the exact type but something like `IsAssignableFrom`.

Not to be silly but I personally think that it is not a great idea to seal classes when others might be extending things on a framework, considering linq2db is a data access framework. Extending from attributes can have variety of reasons from above examples to many others that don't come to our mind now, as well as defining default values or customizing other things.

For instance, I might want to prefix every single table with `ut_` (user table) and define a custom `UserTableAttribute` inheriting from `TableAttribute` and customizing the `tableName` parameter in the constructor before passing it to the base.

But again, I am not trying to push you to unseal classes. Just trying to explain why one might need them unsealed.

I appreciate your comments and replies yet another time. Great work.

**Comment by CoskunSunali** [comment]
@jogibear9988 Yes but that is too much of work for reasons like I mentioned above and is another thing to maintain :)

That is, as long as you have reasons to seal the attributes of course.

**Comment by jogibear9988** [comment]
I think changing all to sealed when our code does not support inherited attributes would be the correct solution (and does not require a major version increase even it‘s a breaking change because inherited attributes then didn‘t work before!)

**Comment by jogibear9988** [comment]
okay, if we support inheritance, i think all should be unsealed

**Comment by jogibear9988** [comment]
im looking to do it!

i‘ve already looked, seems getattributes call supports inherited attributes

**Comment by jogibear9988** [comment]
so changes should be minimal

**Comment by CoskunSunali** [comment]
> i‘ve already looked, seems getattributes call supports inherited attributes

That explains how my `CustomTableAttribute : TableAttribute` worked fine.

I am glad the changes won't require you to do a lot of work.

You guys are of amazing support. Thank you.
